### PR TITLE
Save settings immediately

### PR DESCRIPTION
**Pull request by samhed**
Instead of saving settings when clicking on apply or toggling the settings panel, save them immediately after a change. This is more in line with how most modern interfaces work and saves us from having two different 'states' to keep track of in the UI code (cookie and HTML-elem value).

### CODE REVIEW

**Comment by samhed** [comment]
Another old annoying behavior that is removed with this change is the fact that toggling (by pressing the settings button again) the settings panel did save the settings while closing (by pressing some other toolbar button) didn't.

File: app/ui.js [hunk 1]
```diff
@@ -367,14 +367,41 @@ var UI;
...
+            var settingElem = document.getElementById("noVNC_setting_" + name);
+            if (changeFunc === undefined) {
+                settingElem.addEventListener('change', function () {
+                    UI.saveSetting(name);
+                });
+            } else {
+                settingElem.addEventListener('change', changeFunc);
+            }
+        },
+
```
Comment (CendioOssman) [inline app/ui.js]: If you wish to merge code paths a bit you could do this instead:

```JavaScript
if (changeFunc === undefined) {
    changeFunc = function () { UI.saveSettings(name); };
}
```

File: app/ui.js [hunk 2]
```diff
@@ -367,14 +367,41 @@ var UI;
...
                 .addEventListener('click', UI.toggleSettingsPanel);
-            document.getElementById("noVNC_settings_apply")
-                .addEventListener('click', UI.settingsApply);
 
-            document.getElementById("noVNC_setting_resize")
-                .addEventListener('change', UI.enableDisableViewClip);
+            UI.addSettingChangeHandler('encrypt');
+            UI.addSettingChangeHandler('true_color');
+            if (Util.browserSupportsCursorURIs()) {
+                UI.addSettingChangeHandler('cursor');
```
Comment (CendioOssman) [inline app/ui.js]: Why can't we add this unconditionally? Won't it be disabled where unsupported?

File: app/ui.js [hunk 3]
```diff
@@ -139,9 +139,10 @@ var UI;
                 UI.toggleControlbarSide();
             }
 
-            // Setup and initialize event handlers
+            UI.initFullscreen();
```
Comment (CendioOssman) [inline app/ui.js]: Why the move?

File: app/ui.js [hunk 4]
```diff
@@ -265,6 +262,9 @@ var UI;
             // resize events aren't available for elements
             window.addEventListener('resize', UI.updateControlbarHandle);
 
+            document.getElementById("noVNC_status")
+                .addEventListener('click', UI.hideStatus);
```
Comment (CendioOssman) [inline app/ui.js]: The status isn't really part of the control bar.

File: app/ui.js [hunk 5]
```diff
@@ -502,7 +502,6 @@ var UI;
             }
 
             UI.enableDisableViewClip();
-            document.getElementById('noVNC_setting_resize').disabled = UI.connected;
```
Comment (CendioOssman) [inline app/ui.js]: Isn't more code required? Don't we have to trigger an update of the canvas size and position?

Comment (samhed) [inline app/ui.js]: I'm not sure what you mean? It works just fine

File: app/ui.js (hunk 3 above)
Comment (samhed) [inline app/ui.js]: It isn't really related to setting up event handlers, with the new separation it felt to have more in common with the init functions

File: app/ui.js (hunk 1 above)
Comment (samhed) [inline app/ui.js]: I like that, I'll change

File: app/ui.js (hunk 2 above)
Comment (samhed) [inline app/ui.js]: mostly just did the same thing as the old code here, but you are right, it's unnecessary.

File: app/ui.js (hunk 4 above)
Comment (samhed) [inline app/ui.js]: Yeah.. well it doesn't fit anywhere tbh. But I'll move it into start() I guess

**Comment by kanaka** [comment]
I don't have time to do a code review, but I heartily approve of the general idea.

**Comment by samhed** [comment]
Ok, thank you Joel.

I have rebased on the current master and updated the commits after your comments @CendioOssman, thanks for the feedback!

File: app/ui.js (hunk 5 above)
Comment (CendioOssman) [inline app/ui.js]: It doesn't here. Changing the setting doesn't do anything. I have to resize the window for things to take effect. Also, switching from "scaling" to "none" is just broken.

**Comment by CendioOssman** [comment]
PS. Why don't you have your own fork for your own development? ;)

File: app/ui.js [hunk 6]
```diff
@@ -375,14 +375,38 @@ var UI;
...
-            document.getElementById("noVNC_setting_resize")
-                .addEventListener('change', UI.enableDisableViewClip);
+            UI.addSettingChangeHandler('encrypt');
+            UI.addSettingChangeHandler('true_color');
+            UI.addSettingChangeHandler('cursor');
+            UI.addSettingChangeHandler('resize');
+            UI.addSettingChangeHandler('resize', UI.enableDisableViewClip);
+            UI.addSettingChangeHandler('clip');
+            UI.addSettingChangeHandler('clip', UI.updateViewClip);
+            UI.addSettingChangeHandler('clip', UI.updateViewDrag);
```
Comment (CendioOssman) [inline app/ui.js]: Why are these needed at this point? These settings don't take effect until we are connected, and you can't change those settings once connected.

Comment (samhed) [inline app/ui.js]: You are correct, the extra handlers for 'clip' are not needed. I was half way into a solution where we could change that setting while connected, but I never finished it.

File: app/ui.js (hunk 5 above)
Comment (samhed) [inline app/ui.js]: Yes, you are right.. I'm not sure why I thought that was acceptable tbh. I'll take a stab at fixing these things

**Comment by samhed** [comment]
Changing the resize setting while connected should work as intended now. Thanks for testing!

> PS. Why don't you have your own fork for your own development? ;)

I do, I should probably use that more often :P

**Comment by samhed** [comment]
Had forgotten to set the correct scaling for mouse coordinates, fixed now.

File: app/ui.js [hunk 7]
```diff
@@ -375,14 +375,36 @@ var UI;
...
-            document.getElementById("noVNC_settings_apply")
-                .addEventListener('click', UI.settingsApply);
 
-            document.getElementById("noVNC_setting_resize")
-                .addEventListener('change', UI.enableDisableViewClip);
+            UI.addSettingChangeHandler('encrypt');
+            UI.addSettingChangeHandler('true_color');
+            UI.addSettingChangeHandler('cursor');
+            UI.addSettingChangeHandler('resize');
+            UI.addSettingChangeHandler('resize', UI.enableDisableViewClip);
```
Comment (CendioOssman) [inline app/ui.js]: This should be removed as well, I assume?

Comment (samhed) [inline app/ui.js]: We will actually need it I believe, since choosing any of the two scaling settings forces clipping to be disabled

Comment (CendioOssman) [inline app/ui.js]: Alright. But it should still be in the patch that makes it possible to change when the session is active in that case.

Comment (samhed) [inline app/ui.js]: Ah, I'll move it to the correct commit

Comment (samhed) [inline app/ui.js]: Done
